src/tls_stream.rs

```diff
@@ -193,62 +231,57 @@ impl<S> TlsStream<S>
-            let mut inbuf_desc = winapi::SecBufferDesc {
-                ulVersion: winapi::SECBUFFER_VERSION,
-                cBuffers: 2,
-                pBuffers: inbufs.as_mut_ptr(),
-            };
-
-            let outbufs = &mut [winapi::SecBuffer {
-                                    cbBuffer: 0,
-                                    BufferType: winapi::SECBUFFER_TOKEN,
-                                    pvBuffer: ptr::null_mut(),
-                                },
-                                winapi::SecBuffer {
-                                    cbBuffer: 0,
-                                    BufferType: winapi::SECBUFFER_ALERT,
-                                    pvBuffer: ptr::null_mut(),
-                                },
-                                winapi::SecBuffer {
-                                    cbBuffer: 0,
-                                    BufferType: winapi::SECBUFFER_EMPTY,
-                                    pvBuffer: ptr::null_mut(),
-                                }];
-            let mut outbuf_desc = winapi::SecBufferDesc {
-                ulVersion: winapi::SECBUFFER_VERSION,
-                cBuffers: 3,
-                pBuffers: outbufs.as_mut_ptr(),
-            };
+            let pos = self.enc_in.position() as usize;
+            let mut inbufs = [secbuf(winapi::SECBUFFER_TOKEN,
+                                     Some(&mut self.enc_in.get_mut()[..pos])),
+                              secbuf(winapi::SECBUFFER_EMPTY, None)];
+            let mut inbuf_desc = secbuf_desc(&mut inbufs);
+            debug!("initializing/accepting with {} input bytes", pos);
+
+            let mut outbufs = [secbuf(winapi::SECBUFFER_TOKEN, None),
+                               secbuf(winapi::SECBUFFER_ALERT, None),
+                               secbuf(winapi::SECBUFFER_EMPTY, None)];
+            let mut outbuf_desc = secbuf_desc(&mut outbufs);
 
             let mut attributes = 0;
 
-            let status = secur32::InitializeSecurityContextW(self.cred.get_mut(),
-                                                             self.context.get_mut(),
-                                                             domain,
-                                                             INIT_REQUESTS,
-                                                             0,
-                                                             0,
-                                                             &mut inbuf_desc,
-                                                             0,
-                                                             ptr::null_mut(),
-                                                             &mut outbuf_desc,
-                                                             &mut attributes,
-                                                             ptr::null_mut());
+            let status = if self.accept {
+                let ptr = if self.accept_first {
```

Comment:
Is it possible to avoid this weirdness and always return an context from the initial call to `initialize`? Is it happy if it's passed a 0-byte input buffer?